src/sentry/api/serializers/rest_framework/widget.py

```diff
@@ -0,0 +1,47 @@
+from __future__ import absolute_import
+
+from rest_framework import serializers
+
+from sentry.api.bases.discoversavedquery import DiscoverSavedQuerySerializer
+from sentry.api.serializers.rest_framework import JSONField, ListField, ValidationError
+from sentry.models import WidgetDisplayTypes, WidgetDataSourceTypes
+
+
+class WidgetDataSourceSerializer(serializers.Serializer):
+    name = serializers.CharField(required=True)
+    data = JSONField(required=True)
+    type = serializers.CharField(required=True)
+    order = serializers.IntegerField(required=True)
+
+    def validate_type(self, attrs, source):
+        type = attrs[source]
+        if type not in WidgetDataSourceTypes.TYPE_NAMES:
+            raise ValidationError('Widget data source type %s not recognized.' % type)
+        attrs[source] = WidgetDataSourceTypes.get_id_for_type_name(type)
+        return attrs
+
+    def validate(self, data):
+        if data['type'] == WidgetDataSourceTypes.DISCOVER_SAVED_SEARCH:
```

Comment:
Does the `data` attribute need validation if it is not this type?

Reply:
Yes, but right now that's the only type we have; it'll likely be expanded in the future, but for now that's the only one. I can raise an error if it doesn't have the type given.

Reply:
An error is better imo. Accepting the data could leave us with invalid state in the database and that will require manual intervention to clean up/remove.